src/usrp_drivers/usrp_driver.cpp

```diff
@@ -63,6 +61,131 @@ static clocks_t borealis_clocks;
 // crash to avoid writing corrupt metadata to file.
 #define MAX_CLOCK_DRIFT 1000.0  // seconds
 
+/// Struct replacing the driver_packet.proto definition. This will be populated
+/// from the messages received from radar_control.
+typedef struct {
+  uint32_t sequence_num{};
+  double rxrate{};
+  double txrate{};
+  double txcenterfreq{};
+  double rxcenterfreq{};
+  uint32_t num_rx_samps{};
+  uint32_t num_tx_samps{};
+  double seqtime{};
+  double sample_timing{};
+  bool burst_start{};
+  bool burst_end{};
+  bool align_sequences{};
+  uint32_t buffer_offset{};
+} RadctrlPacket;
+
+/**
+ * @brief      Parses a string into a RadctrlPacket, if possible.
+ *
+ * @param[in]  message  The message in string form
+ *
+ * @return     A struct containing the parsed message.
+ */
+RadctrlPacket parse_radctrl_packet(std::string message) {
+  std::istringstream iss(message);
+  RadctrlPacket packet;
+
+  std::string field_name;
+  std::string token;
+
+  while (iss >> token) {
+    // Now token contains a `name=value` string, which we want to parse
+    auto idx = token.find("=");
+    std::string key = token.substr(0, idx);
+    std::stringstream val(token.substr(idx + 1, token.length()));
+
+    if (key == "sequence_num") {
+      val >> packet.sequence_num;
+    } else if (key == "rxrate") {
+      val >> packet.rxrate;
+    } else if (key == "txrate") {
+      val >> packet.txrate;
+    } else if (key == "txcenterfreq") {
+      val >> packet.txcenterfreq;
+    } else if (key == "rxcenterfreq") {
+      val >> packet.rxcenterfreq;
+    } else if (key == "num_rx_samps") {
+      val >> packet.num_rx_samps;
+    } else if (key == "num_tx_samps") {
+      val >> packet.num_tx_samps;
+    } else if (key == "seqtime") {
+      val >> packet.seqtime;
+    } else if (key == "sample_timing") {
+      val >> packet.sample_timing;
+    } else if (key == "burst_start") {
+      val >> packet.burst_start;
+    } else if (key == "burst_end") {
+      val >> packet.burst_end;
+    } else if (key == "align_sequences") {
+      val >> packet.align_sequences;
+    } else if (key == "buffer_offset") {
+      val >> packet.buffer_offset;
+    } else {
+      RUNTIME_MSG("Invalid key `" << token
+                                  << "` in message from radar_control");
+      std::this_thread::sleep_for(std::chrono::seconds(1));
+      exit(1);
+    }
+  }
+  return packet;
+}
+
+/// Struct replacing the rxsamplesmetadata.proto definition. This will be
+/// populated and converted to messages to send to
+/// rx_signal_processing.
+typedef struct {
```

Comment:
Why do the variable definitions in this struct not have the `{};` at the end of each name, like it is done in `RadCtrlPacket`?

Reply:
The `{}` gives the field a default value, which we don't need to do for `RxPacket` as it gets fully populated all at once. With `RadctrlPacket`, the pre-commit `cpp-check` tool flagged the lack of default values and wouldn't let me commit without it.

Reply:
Gotcha, good to know.